SyncTests/BatchingClientTests.swift

```diff
@@ -0,0 +1,257 @@
+            return deferMaybe(self.emptyResponse)
+        }
+
+        let onCollectionUpload: (POSTResult -> Void) = { _ in onCollectionUploadCalled = true }
+
+        let batch = Sync15BatchClient(config: miniConfig, ifUnmodifiedSince: nil, serializeRecord: serializeRecord, uploader: uploader)
+        batch.addRecords([])
+        batch.commit(onCollectionUpload).succeeded()
+
+        // Shouldn't invoke the callback if we didn't actually upload anything
+        XCTAssertFalse(onCollectionUploadCalled)
+        XCTAssertFalse(onUploadCalled)
+    }
+
+    func testSinglePOSTUpload() {
+        var collectionUploadCount = 0
+        var uploadOpCount = 0
+
+        let uploader: BatchUploadFunction = { lines, timestamp, queryParams in
+            // Single POST should not have query parameters attached
+            XCTAssertNil(queryParams)
+            uploadOpCount += 1
+            return deferMaybe(self.emptyResponse)
+        }
+
+        let onCollectionUpload: (POSTResult -> Void) = { _ in collectionUploadCount += 1 }
+
+        let batch = Sync15BatchClient(config: miniConfig, ifUnmodifiedSince: nil, serializeRecord: serializeRecord, uploader: uploader)
+        let records = generateMockRecords(miniConfig.maxPostRecords)
+        batch.addRecords(records)
+        batch.commit(onCollectionUpload).succeeded()
+
+        // Should only have called the upload/collection callbacks once
+        XCTAssertEqual(collectionUploadCount, 1)
+        XCTAssertEqual(uploadOpCount, 1)
+    }
+
+
+    func testBatchUpload() {
+        var collectionUploadCount = 0
+        var uploadOpCount = 0
+
+        var startedBatch: Bool = false
+        var committedBatch: Bool = false
+
+        let uploader: BatchUploadFunction = { lines, timestamp, queryParams in
+            uploadOpCount += 1
+
+            if let params = queryParams where params.contains({ $0.name == "batch" && $0.value == "true" }) {
+                startedBatch = true
+                let batchStart = POSTResult(modified: 100000, success: [], failed: [:], batchToken: "token")
+                return deferMaybe(StorageResponse(value: batchStart, metadata: ResponseMetadata(status: 200, headers: [:])))
+            }
+
+            if let params = queryParams where params.contains({ $0.name == "commit" && $0.value == "true" }) {
+                committedBatch = true
+            }
+
+            return deferMaybe(self.emptyResponse)
+        }
+
+        let onCollectionUpload: (POSTResult -> Void) = { _ in collectionUploadCount += 1 }
+
+        let batch = Sync15BatchClient(config: miniConfig, ifUnmodifiedSince: nil, serializeRecord: serializeRecord, uploader: uploader)
+        let records = generateMockRecords(miniConfig.maxTotalRecords)
+        batch.addRecords(records)
+        batch.commit(onCollectionUpload).succeeded()
+
+        // Should only have called the upload/collection callbacks once
+        XCTAssertEqual(collectionUploadCount, 1)
```

Comment:
Same. Check that we have the right number of records. (Extra points for being thorough and making sure they're the right records, and not just the same payload over and over!)